src/nools/lib.js

```diff
@@ -194,7 +194,15 @@ function emitTargetFor(targetConfig, c, r) {
   var instanceDoc = isEmittingForReport ? r : c.contact;
   var instance = createTargetInstance(targetConfig.id, instanceDoc, pass);
 
-  instance._id = (targetConfig.idType === 'report' ? r && r._id : c.contact._id) + '~' + targetConfig.id;
+  var instanceId;
+  if (typeof targetConfig.idType === 'function') {
+    instanceId = targetConfig.idType(c, r);
+  } else if (targetConfig.idType === 'report' && r) {
+    instanceId = r._id;
+  } else {
+    instanceId = c.contact._id;
+  }
```

Comment:
This differs a little from the previous logic, example scenario: 
`targetConfig = { appliesTo: 'contacts', idType: 'report', id: 'target' }`
`c.contact = { _id: 'contact' }` `r = undefined`
I assume that, when appliesTo is `'contacts'`, `r` is not defined, the current code would generate this id: `~target`. New code would generate `contact~target`. 
Having a feeling the new is more correct? 

The only comment I would have here is that we could maybe move this ID generation logic in the `createTargetInstance` function below.

Reply:
Thanks! I've moved this (mostly) into `createTargetInstance`.

Also.. I'm not totally sure I understand the implications of the change you pointed out. I'm thinking I shouldn't change it. Last time, I made a fairly benign change to the IDs and it really had a huge fallout in how some projects were unit tested. I've reverted that change.